PlatformUI Context

miguelcleverti opened the pull request:
### Reworked PlatformUI BDD implementation

Removed JS completely and used the proper Mink methods to do the same thing. FieldTypes sentences (as seen in ezsystems/PlatformUIBundle#211) base implementations are also done in conjunction with ezsystems/BehatBundle#24

bdunogier commented on Features/Context/PlatformUI.php:
```diff
@@ -0,0 +1,191 @@
+
+class PlatformUI extends Context
+{
+    const USER = "admin";
```
We should make that configurable somehow. It is quite convenient to run BDD tests on an existing instance without changing your login & password.

bdunogier commented on Features/Context/PlatformUI.php:
```diff
@@ -0,0 +1,191 @@
+    private $platformStatus = self::NOT_WAITING;
+
+    /**
+     * Waits for Javascript to finnish by checking the loading tags of the page
```
"finish", one 'n'

bdunogier commented:
I'm a bit lost here: are the FieldType scenarii the same than then ones from https://github.com/ezsystems/PlatformUIBundle/pull/211 ? I can't see common commits between both pull requests.

andrerom commented:
@miguelcleverti why is there a integer.feature here and also another one in #234 ?

miguelcleverti commented:
@bdunogier / @andrerom I added the scenario by mistake in this branch, I removed it.
@bdunogier now you can pass the username/password as a parameter in behat.yml context configuration, but it still defaults to admin/publish if nothing is provided are provided

bdunogier commented:
> @bdunogier now you can pass the username/password as a parameter in behat.yml context configuration, but it still defaults to admin/publish if nothing is provided are provided

It makes sense, but wouldn't it be more consistent if the default login/password was in behat.yml.dist ? That way people know immediately how to change it.

joaoinacio commented:
@bdunogier: do you mean `behat.yml`? (see https://github.com/ezsystems/ezpublish-community/pull/207 ;) )

andrerom commented:
Seems @lolautruche changed CS for all repos no?

lolautruche commented:
@andrerom only for PlatformUIBundle. But yes, I changed the script on Jenkins

miguelcleverti commented:
@lolautruche @andrerom some of the files are not related to my PR, should I change only my files or do you want me to fix everything?

lolautruche commented:
@miguelcleverti Don't change anything for now.
We're gonna switch to PSR-2 for CS, so you'll have to change this, but wait for #239 to be merged

lolautruche commented:
@miguelcleverti #239 is now merged. Please rebase your branch now and update your PR to comply PSR-2.

Thanks :-)

miguelcleverti commented:
Changed the cs and squashed some commits

andrerom commented:
looks better now, your pov @bdunogier ?

bdunogier commented on Features/Context/PlatformUI.php:
```diff
@@ -0,0 +1,198 @@
+
+/**
+ * File containing the main context class for PlatformUI
+ * This flie contains the mapping of the BDD sentences to the functions that implement them
```
flie => fly. Bugfree code ;-)à

bdunogier commented on Features/Context/PlatformUI.php:
```diff
@@ -0,0 +1,198 @@
+<?php
+
+/**
+ * File containing the main context class for PlatformUI
```
I'd stick to the standard header ("This file is part of the EzSystemsBehatBundle package."), and document the class in the class's phpdoc (it is useful, and it will make it more visible in IDEs).

dpobel commented on Features/Standard/platformui.feature:
```diff
@@ -0,0 +1,132 @@
+        Given I go to PlatformUI app with username "admin" and password "publish"
+        And I click on "Root location" link
+        When I click on the tab "Activity"
+        Then I should see "Activity content" text
```
useless scenario, the activity tab is a manual testing thing, not a feature of the application

dpobel commented on Features/Standard/platformui.feature:
```diff
@@ -0,0 +1,132 @@
+    Scenario: As a admin User, I want to see navigate menu
+        Given I go to PlatformUI app with username "admin" and password "publish"
+        And I click on the menu zone "Create"
+        When I click on the sub-menu option "Insite editing"
```
the *Insite editing* menu is gone for quite while now and was hardcoded like the activity tab to manually tests some behaviour.

dpobel commented on Features/Standard/platformui.feature:
```diff
@@ -0,0 +1,132 @@
+        Then I should see "Site 1" text
+
+    @javascript
+    Scenario: As a admin User, I want to see side menu
```
not *side menu*, but *action bar*

dpobel commented on Features/Standard/platformui.feature:
```diff
@@ -0,0 +1,132 @@
+    @javascript
+    Scenario: As a admin User, I want to see side menu
+        Given I go to PlatformUI app with username "admin" and password "publish"
+        And I click on the menu zone "Create"
```
not *menu zone* but *button action*

dpobel commented on Features/Standard/platformui.feature:
```diff
@@ -0,0 +1,132 @@
+    Scenario: As a admin User, I want to see side menu
+        Given I go to PlatformUI app with username "admin" and password "publish"
+        And I click on the menu zone "Create"
+        And I click on the "Content structure" link
```
it's not just a link, it's a "Navigation item"

dpobel commented on Features/Standard/platformui.feature:
```diff
@@ -0,0 +1,132 @@
+        Given I go to PlatformUI app with username "admin" and password "publish"
+        And I click on the menu zone "Create"
+        And I click on the "Content structure" link
+        And I click on the side menu option "Content tree"
```
not "side menu option" but "button action"

dpobel commented:
The scenario are very confusing because you call almost everything *side menu*, *navigation menu* or worse *link*. Widgets have names, it would be nice to try to stick to them, here's schema to help:

![platformui_domain](https://cloud.githubusercontent.com/assets/305563/7987855/72da8b46-0ae0-11e5-9c1f-137362a77ad2.png)

bdunogier commented on Features/Standard/platformui.feature:
```diff
@@ -0,0 +1,132 @@
+Feature: Login on PlatformUI
```
Is it really what is tested here ? Besides the first scenario, they look more like navigation scenarios.

andrerom commented:
Awesome @dpobel, @ezsystems/qa-team can the domain language be updated to use the terms described by @dpobel ?

bdunogier commented:
Ping @miguelcleverti.

miguelcleverti commented:
@bdunogier, @dpobel, @andrerom the changes to the domain language that @dpobel referred were mostly already done, the standard feature file was outdated. I improved the bar menus button clicking, now there is a difference between the discovery, the action and the edit menu bar. I also rebased to master.

joaoinacio commented on Features/Context/SubContext/CommonActions.php:
```diff
@@ -0,0 +1,223 @@
+
+    /**
+     * @Given I click (on) the edit action bar button :button
+     * Click on a PlatformUI action bar
```
Nitpick: should probably specify the "Edit action bar"

andrerom commented on Features/Context/SubContext/CommonActions.php:
```diff
@@ -0,0 +1,223 @@
+     *
+     * @param  string   $button     Text of the element to click
+     */
+    public function overNavigationZone($zone)
```
Maybe I'm missing something here, but the hovering behavior has been removed right? /cc @dpobel

miguelcleverti replied:
This is a piece of sneaky code that survived the clean up :sweat_smile:. Removed!

andrerom commented on Features/Context/SubContext/Fields.php:
```diff
@@ -0,0 +1,206 @@
+trait Fields
+{
+    /**
+     * @When I create a content of this type
```
what is "this" referring to? is this a type defined by earlier sentence?

miguelcleverti replied:
@andrerom That is right. This sentence is to be used in conjunction with the other ones for creating contenttypes in way that is easier to read/understand. These sentences are more focused on the fieldtypes so they provide an abstraction for creating contenttypes and content.

bdunogier commented:
During the workshops, it was asked if there was like a minimal behat test running on PlatformUI. The answer was unfortunately no.

Seeing this PR, it seems to me that we have like everything we need or close. If we extract the Authentication Context + the Common one, we can have a simple scenario that tries to login on PlatformUI, and asserts that the homepage is how it should be. 

This would ensure that PlatformUI _works_ (e.g. you can login, and the basic content is retrieved) on pull-requests. It's a small step, bug a big insurance.

Could we do that ? Ping @andrerom @miguelcleverti @dpobel.

bdunogier commented:
Alternative is to merge it, since it passes, and update the sentences as we use them. I'm not a huge fan* of committing code we don't use and agreed upon, but I could agree.

![Turbine](http://imgs.xkcd.com/comics/turbine.png)

andrerom commented:
+1

bdunogier commented:
+1

bdunogier commented:
Merged, please rebase issues that depend on this.
